File: PythonCode/ChapterNH/Exercise_8_2_2.py
"""
 This script plots the determinant of a coefficient matrix of a set 
 of equations which determines the resonance energies for a potential 
 consisting of two rectangular barriers. 

 Inputs are
 For the barriers:
 V0 - The height of each barier (it is the same for both)
 d - half the distance between the barriers
 w - the width of each barrier.

 For the energy ranges:
 ImEMax - the maximal absolute value of negative imaginary energy
 dE - resolution in the enegy grid it is same for the real and the 
 imaginary part.
"""
# Libraries
import numpy as np
from matplotlib import pyplot as plt

# Physical input parameters
V0 = 4     
d = 3     
w = .5     

# Energy grid
ReEMax = V0
ImEMax = .25            # Assign the absolute value of maximally negative
dE = .01

# Vectors of real and imaginary energy
ReEvector = np.arange(0, ReEMax, dE)
ImEvector = np.arange(-ImEMax, 0, dE)

# Alloacte
DetMatSymm = np.zeros((len(ReEvector), len(ImEvector)))
DetMatAntiSymm = np.zeros((len(ReEvector), len(ImEvector)))

# Define parmeters, onset and offset of the barriers
dPlus = d + w/2
dMinus = d - w/2


# Construct matrices with values of determinants
# Symmetric case
ReInd=0
for realE in ReEvector:
  ImInd = 0
  for imagE in ImEvector:
    # Energy-dependent parameters
    E = realE + 1j*imagE            # Complex energy
    kappa = np.sqrt(2*(V0-E))       
    k = np.sqrt(2*E)
    
    # Coefficient matrix
    Mat=[[np.cos(k*dMinus), -np.exp(-kappa*dMinus), -np.exp(kappa*dMinus), 0],
    [-k*np.sin(k*dMinus), kappa*np.exp(-kappa*dMinus), 
                             -kappa*np.exp(kappa*dMinus), 0],
    [0, np.exp(-kappa*dPlus), np.exp(kappa*dPlus), -np.exp(1j*k*dPlus)],
    [0, -kappa*np.exp(-kappa*dPlus), kappa*np.exp(kappa*dPlus), 
                             -1j*k*np.exp(1j*k*dPlus)]]
    
    # Calculate and assign determinant (in absolute value)
    DetMatSymm[ReInd,ImInd]=np.abs(np.linalg.det(Mat))
    ImInd = ImInd + 1              # Update index
  ReInd = ReInd + 1                # Update index


# Construct matrices with values of determinants
# Antisymmetric case
ReInd=0
for realE in ReEvector:
  ImInd = 0
  for imagE in ImEvector:
    # Energy-dependent parameters
    E = realE + 1j*imagE            # Complex energy
    kappa = np.sqrt(2*(V0-E))       
    k = np.sqrt(2*E)
    
    # Coefficient matrix
    Mat=[[np.sin(k*dMinus), -np.exp(-kappa*dMinus), -np.exp(kappa*dMinus), 0],
    [k*np.cos(k*dMinus), kappa*np.exp(-kappa*dMinus), 
                             -kappa*np.exp(kappa*dMinus), 0],
    [0, np.exp(-kappa*dPlus), np.exp(kappa*dPlus), -np.exp(1j*k*dPlus)],
    [0, -kappa*np.exp(-kappa*dPlus), kappa*np.exp(kappa*dPlus), 
                             -1j*k*np.exp(1j*k*dPlus)]]
    
    # Calculate and assign determinant (in absolute value)
    DetMatAntiSymm[ReInd,ImInd]=np.abs(np.linalg.det(Mat))
    ImInd = ImInd + 1              # Update index
  ReInd = ReInd + 1                # Update index


# Plot determinants - for both symmetric and antisymmetric case
fig, (ax1, ax2) = plt.subplots(2, 1)
# Symmetric case
ax1.pcolor(ReEvector, ImEvector, np.log(DetMatSymm.T))
ax1.set_xlabel(r'Re  $\varepsilon$')
ax1.set_ylabel(r'Im  $\varepsilon$')
# Anti-symmetric case
ax2.pcolor(ReEvector, ImEvector, np.log(DetMatAntiSymm.T))
ax2.set_xlabel(r'Re  $\varepsilon$')
ax2.set_ylabel(r'Im  $\varepsilon$')
plt.show()

File: MatlabCode/Chapter8_NonHermitian/Exercise_8_2_2.m
% This script plots the determinant of a coefficient matrix of a set 
% of equations which determines the resonance energies for a potential 
% consisting of two rectangular barriers. 
%
% Inputs are
% For the barriers:
% V0 - The height of each barier (it is the same for both)
% d - half the distance between the barriers
% w - the width of each barrier.
%
% For the energy ranges:
% ImEMax - the maximal absolute value of the negative imaginary energy
% dE - resolution in the enegy grid; it is same for the real and the 
% imaginary part.

% Physical input parameters
V0 = 4;     
d = 3;     
w = .5;     

% Energy grid
ReEMax = V0;
ImEMax = 0.25;          % Maximal value of negative imaginary energy
dE = .005;

% Vectors of real and imaginary energy
ReEvector = 0:dE:ReEMax;
ImEvector = -ImEMax:dE:0;

% Allocate
DetMatSymm = zeros(length(ReEvector), length(ImEvector));
DetMatAntiSymm = zeros(length(ReEvector), length(ImEvector));

% Define parmeters, onset and offset of the barriers
dPlus = d + w/2;
dMinus = d - w/2;


% Construct matrices with values of determinants
% Symmetric case
ReInd = 1;
for realE=ReEvector
  ImInd = 1;
  for imagE = ImEvector  
    % Energy-dependent parameters
    E = realE+i*imagE;            % Complex energy
    kappa = sqrt(2*(V0-E));       
    k = sqrt(2*E);
    
    % Coefficient matrix
    Mat = [cos(k*dMinus) -exp(-kappa*dMinus) -exp(kappa*dMinus) 0
    -k*sin(k*dMinus) kappa*exp(-kappa*dMinus) -kappa*exp(kappa*dMinus) 0
    0 exp(-kappa*dPlus) exp(kappa*dPlus) -exp(i*k*dPlus)
    0 -kappa*exp(-kappa*dPlus) kappa*exp(kappa*dPlus) -i*k*exp(i*k*dPlus)];
    
    % Calculate and assign determinant (in absolute value)
    DetMatSymm(ReInd,ImInd) = abs(det(Mat));
    ImInd = ImInd + 1;              % Update index
  end
  ReInd = ReInd + 1;                % Update index
end

% Plot logarithm of |det(A)| for the symmetric case
figure(1)
subplot(2, 1, 1)
% Plot the logarithm - to highlight the minima
pcolor(ReEvector, ImEvector, log(DetMatSymm).')
shading interp
set(gca,'fontsize',15)
xlabel('Re E')
ylabel('Im E')

% Anti-symmetric case - identical except for element 1,1 and 2,1 in the 
% coefficient matrix
ReInd = 1;
for realE = ReEvector
  ImInd = 1;
  for imagE = ImEvector  
    % Energy-dependent parameters
    E = realE+i*imagE;            % Complex energy
    kappa = sqrt(2*(V0-E));       
    k = sqrt(2*E);
    
    % Coefficient matrix
    Mat = [sin(k*dMinus) -exp(-kappa*dMinus) -exp(kappa*dMinus) 0
    k*cos(k*dMinus) kappa*exp(-kappa*dMinus) -kappa*exp(kappa*dMinus) 0
    0 exp(-kappa*dPlus) exp(kappa*dPlus) -exp(i*k*dPlus)
    0 -kappa*exp(-kappa*dPlus) kappa*exp(kappa*dPlus) -i*k*exp(i*k*dPlus)];
    
    % Calculate and assign determinant (in absolute value)
    DetMatAntiSymm(ReInd,ImInd) = abs(det(Mat));
    ImInd = ImInd + 1;              % Update index
  end
  ReInd = ReInd + 1;                % Update index
end

% Plot |det(A)| for the antisymmetric/odd case
subplot(2, 1, 2)
% Plot the logarithm - to highlight the minima
pcolor(ReEvector, ImEvector, log(DetMatAntiSymm).')
shading interp
set(gca,'fontsize',15)
xlabel('Re E')
ylabel('Im E')